google/cloud/bigtable/data/_metrics/data_model.py

```diff
@@ -0,0 +1,514 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+
+from typing import Callable, Any, Generator, Tuple, cast, TYPE_CHECKING
+
+import datetime
+import time
+import os
+import re
+import logging
+
+from enum import Enum
+from functools import lru_cache
+from dataclasses import dataclass
+from dataclasses import field
+from grpc import StatusCode
+
+import google.cloud.bigtable.data.exceptions as bt_exceptions
+
+if TYPE_CHECKING:
+    from google.cloud.bigtable.data._metrics.handlers._base import MetricsHandler
+
+
+# by default, exceptions in the metrics system are logged,
+# but enabling this flag causes them to be raised instead
+ALLOW_METRIC_EXCEPTIONS = os.getenv("BIGTABLE_METRICS_EXCEPTIONS", False)
+LOGGER = (
+    logging.getLogger(__name__) if os.getenv("BIGTABLE_METRICS_LOGS", False) else None
+)
+
+# default values for zone and cluster data, if not captured
+DEFAULT_ZONE = "global"
+DEFAULT_CLUSTER_ID = "unspecified"
+
+# keys for parsing metadata blobs
+BIGTABLE_METADATA_KEY = "x-goog-ext-425905942-bin"
+SERVER_TIMING_METADATA_KEY = "server-timing"
+SERVER_TIMING_REGEX = re.compile(r"gfet4t7; dur=(\d+)")
+
+INVALID_STATE_ERROR = "Invalid state for {}: {}"
+
+
+@dataclass(frozen=True)
+class TimeTuple:
+    """
+    Tuple that holds both the utc timestamp to record with the metrics, and a
+    monotonic timestamp for calculating durations. The monotonic timestamp is
+    preferred for calculations because it is resilient to clock changes, eg DST
+    """
+    utc: datetime.datetime = field(
+        default_factory=lambda: datetime.datetime.now(datetime.timezone.utc)
+    )
+    monotonic: float = field(default_factory=time.monotonic)
+
+
+class OperationType(Enum):
+    """Enum for the type of operation being performed."""
+
+    READ_ROWS = "ReadRows"
```

Comment:
I guess this is a question for Igor. In java we tag ReadRow and ReadRows differently, I wonder if we should do the same for python? Or leave it in sync with the server metrics?

Reply:
I thought that's essentially what the streaming flag was for?

Reply:
I think java ReadRow tag was my mistake :( 
Lets keep this inline with server metrics. I think the correct path forward will be stanley's request profiles